crates/ruma-api/src/lib.rs

```diff
@@ -591,3 +595,82 @@ impl MatrixVersion {
         }
     }
 }
+
+/// A specifier for what path variant of an endpoint to request.
+#[derive(Clone, PartialEq, Eq, Debug)]
+#[allow(clippy::exhaustive_enums)]
+pub enum EndpointPath {
+    /// Select an appropriate path version for a particular matrix version.
+    ///
+    /// Note: When using
+    /// [`try_into_http_request`](OutgoingRequest::try_into_http_request) with this variant, it can
+    /// emit additional errors;
+    /// - [`NoUnstablePath`](IntoHttpError::NoUnstablePath), if the matrix version is too new, and
+    ///   the endpoint has no defined `unstable` path (for maintenance or legacy reasons.)
+    /// - [`EndpointRemoved`](IntoHttpError::EndpointRemoved), if the matrix version is too old,
+    ///   and the endpoint is known to be removed in the same or a newer version.
+    ForMatrixVersion(MatrixVersion),
+
+    /// Select any path, preferring stable.
+    ///
+    /// Note: This does not emit any errors like
+    /// [`ForMatrixVersion`](EndpointPath::ForMatrixVersion) does, if the server's lowest matrix
+    /// version has removed the endpoint, or it has not yet stabilized the endpoint, this may
+    /// result in runtime HTTP errors, it is highly recommended to fetch `/versions`, extract
+    /// Ruma-known matrix versions, and pass the preferred version through `ForMatrixVersion`.
+    PreferStable,
+}
+
+impl From<MatrixVersion> for EndpointPath {
+    fn from(ver: MatrixVersion) -> Self {
+        Self::ForMatrixVersion(ver)
+    }
+}
+
+// This function needs to be public, yet hidden, as all `try_into_http_request`s would be using it.
+// We're also using MatrixVersion::repr() here, which is crate-private.
+#[doc(hidden)]
+pub fn select_path<'a>(
+    path: EndpointPath,
+    metadata: &'_ Metadata,
+    unstable: Option<fmt::Arguments<'a>>,
+    r0: Option<fmt::Arguments<'a>>,
+    stable: Option<fmt::Arguments<'a>>,
+) -> Result<fmt::Arguments<'a>, IntoHttpError> {
+    let version = match path {
+        EndpointPath::ForMatrixVersion(v) => v,
+        EndpointPath::PreferStable => {
+            return Ok(stable.or(r0).or(unstable).expect("one of three paths to be defined"))
+        }
+    };
+
+    if let Some(removed_ver) = metadata.removed {
+        if version.repr() >= removed_ver.repr() {
```

Comment:
Todo, after https://github.com/ruma/ruma/pull/872: change these to `is_superset_of`

Reply:
We can probably also inline this bit for the macro, and remove this function entirely.

Reply:
I like that this function exists outside the macro. It should reduce code bloat a bit, since it has no type or const generics.